c7n/resources/dms.py

```diff
@@ -105,3 +121,114 @@ def process(self, resources):
         client = local_session(self.manager.session_factory).client('dms')
         for arn, r in zip(self.manager.get_arns(resources), resources):
             client.delete_replication_instance(ReplicationInstanceArn=arn)
+
+
+@ReplicationInstance.action_registry.register('tag')
+class InstanceTag(Tag):
+    """
+    Add tag(s) to a replication instance
+
+    :example:
+
+        .. code-block:: yaml
+
+            policies:
+                - name: tag-dms-required
+                  resource: dms-instance
+                  filters:
+                    - "tag:RequireTag": absent
+                  actions:
+                    - type: tag
+                      key: RequiredTag
+                      value: RequiredTagValue
+    """
+    permissions = ('dms:AddTagsToResource',)
+
+    def process_resource_set(self, resources, tags):
+        client = local_session(self.manager.session_factory).client('dms')
+        tags_list = []
+        for t in tags:
+            tags_list.append({'Key': t['Key'], 'Value': t['Value']})
+        for r in resources:
+            try:
+                client.add_tags_to_resource(
+                    ResourceArn=r['ReplicationInstanceArn'],
+                    Tags=tags_list)
+            except ClientError as e:
+                self.log.exception(
```

Comment:
i'd suggest seriously considering that blindly catchings errors is not a good practice, and that ideally we should only be catching explicit known errors and letting the rest bubble through.  alternatively .. designing an error capture and reporting sys with serialized outputs alongside resources.json.

Reply:
So only use try/except in cases where we know the errors? ie: Delete action catch for Resource not found?

Reply:
In this and other cases such as S3, i've been trying to push an error handling scheme where don't stop the policy run if one resource throws an error.  We have had a lot of problems not catching errors for the last couple of months and have done a lot of work to catch errors and log instead of throwing.  I think it would be a better practice to process all resources that can be processed, and log errors about resources that could not be processed.  Downstream tasks can be created to alert and handle errors in either log messages or we can handle it differently, like your suggestion with a error resource file.

Reply:
Digging into other patterns, it would be useful to raise if there was an error code that would signify you probably can't access the other resources.  For example, in the case of ECS, a serviceNotFound code would mean you probably won't be able to execute the action on all the other resources and there is no reason to continue so raise.

In this case however, it doesn't matter what error is thrown when trying to add tags for the resource.  We should try to add tags to all the resources before failing the policy.

Reply:
its more about try/catch on exception, but check error against known, if you its unknown or hasn't been reasoned about its really rather hard to say what the behavior should be, and rather then guess i'd rather surface it to the user, rather than have it silently pass. It would be better to adjust the list with known exceptions over time (or reason about that at commit time), rather than just blithely ignoring errors. The primary issue here is that logs are not a good way to convey errors, and that developing a new mechanism to report resource level errors is needed.

Reply:
I would disagree that logs are not a good way to convey errors.  The "first time user" is probably the only person that's going to run this on their laptops or manually from the command line.  From there, users are going to create automated scheduled jobs that execute custodian, at which point, gathering and alerting on logs is going to be the primary place for setting up alerting and monitoring.  Additionally, this stack trace will bubble up to the logs, where it will be evaluated and alerted on.

With all that being said, the main thing i'm advocating for is that we process as many resources as possible.  Not process until error and then stop processing, leaving many resources not acted on that could have been acted on.

I don't think anything elaborate needs to be created to report resource level errors.  Simply logging an exception at the time of the error would be sufficient if the message included all the information needed to understand the exception, including resource information.

Reply:
that assumes quite alot of infrastructure and assumptions about log parsing and monitoring.  i've seen too many cron jobs running where the policy failed to act on any resources the log was filled with errors, and the user thought everything was working. a failed cron execution caught their attention more directly. again the notion here is that we should continue to process resources if we catch known/reasoned errors, but catch alls sprinkled everywhere are not a good practice in any system.

Reply:
At this point, there seems to be some fundamental differences in opinion that should not hold up this PR specifically and can be moved to a different forum.

@joshuaroot please check for the specific error `ResourceNotFoundFault`, log that specifically and continue.  Otherwise raise.